src/Silicon.cpp

```diff
@@ -874,124 +1135,493 @@ namespace galsim {
+                for (int i=polyi1; i <= polyi2; i++) {
+                    // Check whether this pixel has charge on it
+                    double charge = deltaData[(j * stride) + (i * step)];
+
+                    if (charge != 0.0) {
+                        change = true;
+
+                        // Work out corresponding index into distortions array
+                        int dist_index = (((x - i + nxCenter) * _ny) + ((_ny - 1) - (y - j + nyCenter))) * verticalPixelStride() + (verticalPixelStride() - 1);
+                        int index = p * verticalPixelStride() + (verticalPixelStride() - 1);
+
+                        // Loop over boundary points and update them
+                        for (int n=0; n < verticalPixelStride(); ++n, --index, --dist_index) {
+                            verticalBoundaryPointsData[index].x =
+                                double(verticalBoundaryPointsData[index].x) +
+                                verticalDistortionsData[dist_index].x * charge;
+                            verticalBoundaryPointsData[index].y =
+                                double(verticalBoundaryPointsData[index].y) +
+                                verticalDistortionsData[dist_index].y * charge;
+                        }
+                    }
+                }
+            }
 
-                // This isn't atomic -- openmp is handling the reduction for us.
-                addedFlux += flux;
+            // update changed array
+            if (change) {
+                if (x < nx) changedData[(x * ny) + y] = true;
+                if (x > 0)  changedData[((x - 1) * ny) + y] = true;
             }
         }
+
+        Bounds<double>* pixelInnerBoundsData = _pixelInnerBounds.data();
+        Bounds<double>* pixelOuterBoundsData = _pixelOuterBounds.data();
 #ifdef _OPENMP
+#ifndef GALSIM_USE_GPU
+#pragma omp parallel for
+#else
+#pragma omp target teams distribute parallel for
+#endif
+#endif
+        for (size_t k=0; k<nxny; ++k) {
+            if (changedData[k]) {
+	        updatePixelBoundsGPU(nx, ny, k, pixelInnerBoundsData,
+                                     pixelOuterBoundsData,
+                                     horizontalBoundaryPointsData,
+                                     verticalBoundaryPointsData);
+                changedData[k] = false;
+            }
         }
+        
+        // update target from delta and zero delta on GPU
+        // CPU delta is not zeroed but that shouldn't matter
+#ifdef _OPENMP
+#ifndef GALSIM_USE_GPU
+#pragma omp parallel for
+#else
+#pragma omp target teams distribute parallel for
 #endif
-        return addedFlux;
-    }
-
-    template <typename T>
-    void Silicon::update(ImageView<T> target)
-    {
-        updatePixelDistortions(_delta.view());
-        target += _delta;
-        _delta.setZero();
+#endif
+        for (int i = 0; i < imageDataSize; i++) {
+            targetData[i] += deltaData[i];
+            deltaData[i] = 0.0;
+	}
```

Comment:
I assume this should be a call to `addDelta`, right?  Looks like an oversight in the previous code, but it's more important now that addDelta isn't just a one-liner.